### PR TITLE
Improve and extend arithmetic operations

### PR DESCRIPTION
**Pull request by NBorge**

### CODE REVIEW

File: tyda/src/test/scala/com/choreograph/tyda/ExprSpec.scala [hunk 1]
```diff
@@ -148,7 +148,25 @@ class ExprSpec extends AnyFunSuite {
...
+  test("support multiply") {
+    val rNode = ExprNode.Reference[(Int, Int)]()
+    val r = Expr.lift(rNode)
+    val lhs = r._1
+    val rhs = r._2
+    val expr = lhs * rhs
+    assert(expr.node == ExprNode.Multiply(Num[Int], Expr.unlift(lhs), Expr.unlift(rhs)))
   }
 
   test("support quotient") {
```
Comment (shambala-ifmo) [inline tyda/src/test/scala/com/choreograph/tyda/ExprSpec.scala]: shouldn't we also test the behavior with floating point numbers?

File: tyda/src/main/scala/com/choreograph/tyda/ExprApi.scala [hunk 2]
```diff
@@ -500,26 +500,47 @@ trait ExprApi[Expr[T]] {
...
+      * types.
+      */
+    infix def *[I: AsExpr.Of[T]](rhs: I): Expr[T] =
+      lift(ExprNode.Multiply(Num[T], unlift(lhs), unlift(AsExpr(rhs))))
 
     /** Returns the result of truncating division of the left operand by the
       * right operand.
       *
-      * Throws an exception if the divisor is zero.
+      * For floating point types this is true division. Throws an exception if
```
Comment (shambala-ifmo) [inline tyda/src/main/scala/com/choreograph/tyda/ExprApi.scala]: according to Num implementation, it throws an exception for floating point types as well 
`      if rhs == 0 then throw new ArithmeticException("/ by zero") else lhs / rhs`

File: tyda/src/main/scala/com/choreograph/tyda/Num.scala [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+  }
+
+  private[tyda] case object Double extends Primitive[scala.Double] {
+    def plus(lhs: scala.Double, rhs: scala.Double): scala.Double = lhs + rhs
+    def minus(lhs: scala.Double, rhs: scala.Double): scala.Double = lhs - rhs
+    def times(lhs: scala.Double, rhs: scala.Double): scala.Double = lhs * rhs
+    def quot(lhs: scala.Double, rhs: scala.Double): scala.Double =
+      if rhs == 0 then throw new ArithmeticException("/ by zero") else lhs / rhs
+    def negate(t: scala.Double): scala.Double = -t
+  }
```
Comment (shambala-ifmo) [inline tyda/src/main/scala/com/choreograph/tyda/Num.scala]: I think we can do 
```suggestion
given Num[Double] = new Primitive[Double] {
    def plus(lhs: Double, rhs: Double): scala.Double = lhs + rhs
    def minus(lhs: Double, rhs: Double): scala.Double = lhs - rhs
    def times(lhs: Double, rhs: Double): scala.Double = lhs * rhs
    def quot(lhs: Double, rhs: Double): scala.Double =
      if rhs == 0 then throw new ArithmeticException("/ by zero") else lhs / rhs
    def negate(t: Double): Double = -t
  }
```
to make it less boilerplate and avoid name clashes

File: tyda/src/test/scala/com/choreograph/tyda/ExprSpec.scala (hunk 1 above)
Comment (NBorge) [reply tyda/src/test/scala/com/choreograph/tyda/ExprSpec.scala]: Added tests for every type.

File: tyda/src/main/scala/com/choreograph/tyda/ExprApi.scala (hunk 2 above)
Comment (NBorge) [reply tyda/src/main/scala/com/choreograph/tyda/ExprApi.scala]: Yes this was later changed, I've updated all of the descriptions.

File: tyda/src/main/scala/com/choreograph/tyda/Num.scala (hunk 3 above)
Comment (NBorge) [reply tyda/src/main/scala/com/choreograph/tyda/Num.scala]: Updated the Num object to follow this pattern for all types.

File: tyda-sql/src/test/resources/golden/ExprEvaluationSparkSqlGoldenSuite.golden [hunk 4]
```diff
@@ -82,11 +82,27 @@ SELECT unix_micros(t1.value) AS value FROM t1
...
 
+------ Test: add two byte numbers
+SELECT (t1._1 + t1._2) AS value FROM t1
+------ end
+
+------ Test: add two double numbers
+SELECT CASE WHEN (((NOT ((t1._1 = CAST('Infinity' AS DOUBLE)) OR (t1._1 = CAST('-Infinity' AS DOUBLE)))) AND (NOT ((t1._2 = CAST('Infinity' AS DOUBLE)) OR (t1._2 = CAST('-Infinity' AS DOUBLE))))) AND ((NOT isnan((t1._1 + t1._2))) AND (((t1._1 + t1._2) < CAST(-1.7976931348623157E308 AS DOUBLE)) OR ((t1._1 + t1._2) > CAST(1.7976931348623157E308 AS DOUBLE))))) THEN raise_error('Double overflow') ELSE (t1._1 + t1._2) END AS value FROM t1
+------ end
+
+------ Test: add two float numbers
```
Comment (eejbyfeldt) [inline tyda-sql/src/test/resources/golden/ExprEvaluationSparkSqlGoldenSuite.golden:117]: Seems like we are generating complicated code here on both BigQuery and Spark. Should we not choose the behavior so that we generate clean efficient code on at least one backend?

File: tyda-sql/src/main/scala/com/choreograph/tyda/sql/SqlDialect.scala [hunk 5]
```diff
@@ -301,6 +302,12 @@ object SqlDialect {
     case NaNIsLargest
   }
 
+  enum FloatingOverflowChecks {
+    case FloatAndDouble
+    case FloatOnly
```
Comment (eejbyfeldt) [inline tyda-sql/src/main/scala/com/choreograph/tyda/sql/SqlDialect.scala]: On BigQuery Float and Double is backed by the same type. Why do we need to make this distinction here?

Comment (eejbyfeldt) [reply tyda-sql/src/main/scala/com/choreograph/tyda/sql/SqlDialect.scala]: Aa because the maximum value is smaller.

File: tyda-sql/src/main/scala/com/choreograph/tyda/sql/SqlDialect.scala [hunk 6]
```diff
@@ -301,6 +302,12 @@ object SqlDialect {
     case NaNIsLargest
   }
 
+  enum FloatingOverflowChecks {
```
Comment (eejbyfeldt) [inline tyda-sql/src/main/scala/com/choreograph/tyda/sql/SqlDialect.scala]: Most of the other dialect knobs are named after the behavior the backend has not what checks should be added. e.g `FloatingAggregate.NaNIsSmallestAndLargest` means that in that backend NaN is smallest and largest value for min/max aggregates and we should add logic to make that uniform. So if we want to be consistent we should follow that.

**Comment by eejbyfeldt** [comment]
Can we add a description with implementation choices made here and their motivations.

File: tyda-sql/src/test/resources/golden/ExprEvaluationSparkSqlGoldenSuite.golden (hunk 4 above)
Comment (eejbyfeldt) [reply tyda-sql/src/test/resources/golden/ExprEvaluationSparkSqlGoldenSuite.golden:117]: Maybe a description would clear this up.

File: tyda-sql/src/main/scala/com/choreograph/tyda/sql/ExprUnparser.scala [hunk 7]
```diff
@@ -155,6 +156,96 @@ private def exprToSqlExpr[T](fullExpr: ExprNode[T], args: UnparserArgs): Result[
...
+            literalToSqlExpr(-Double.MaxValue, Codec.Double, dialect),
+            literalToSqlExpr(Double.MaxValue, Codec.Double, dialect),
+            "Double overflow",
+            codec
+          )
+
+        case _ => returnedResult
+      }
+
+    def _overflowCheckedFloatingResult(
```
Comment (eejbyfeldt) [inline tyda-sql/src/main/scala/com/choreograph/tyda/sql/ExprUnparser.scala]: Or prefix `_` looks a bit like python.

```suggestion
    private def overflowCheckedFloatingResult(
```

File: tyda-test-suites/src/main/scala/com/choreograph/tyda/testsuites/ExprEvaluationSuiteBase.scala
Comment (eejbyfeldt) [inline tyda-test-suites/src/main/scala/com/choreograph/tyda/testsuites/ExprEvaluationSuiteBase.scala:1]: Feels like we should be able to be able to reduce al lot of the repetition here by using helpers to define the tests. e.g

```
  private def testFloatingpoint[T: Codec: Numeric as numeric: Num: FloatingLimits as limits: Arbitrary as arbitrary: SimpleTypeName as name] = {
    import numeric.given
    given Arbitrary[T] = arbitrary.filter {
      value =>
      value.toDouble < Math.sqrt(value.toDouble) || (value == limits.infinity || -value == limits.infinity)
    }
    testHasSameBehavior[(T, T), T](
      s"multiply two $name numbers",
      {  case Expr(a, b) => a * b },
      _ * _
    )
   ...
  }
```

Would make the multiply tests be possible to define once and use both for floating point and bigquery.

Probably there are also other tests that can be implemented once and resued for all numeric types.

File: tyda/src/main/scala/com/choreograph/tyda/Num.scala [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+  * Note: for custom types where only a subset of operations makes sense (e.g.
+  * only + and - for a Duration-like type) prefer defining dedicated extension
+  * methods rather than providing a Num instance.
+  */
+@implicitNotFound(
+  "Type ${T} does not support numeric operations.\n" +
+    "If ${T} is a type parameter consider adding a context bound: [${T}: Num]\n" +
+    "If ${T} is an opaque type consider adding a given Num[${T}] = Num.<evidence> to the companion object\n"
+)
+sealed trait Num[T] {
```
Comment (eejbyfeldt) [inline tyda/src/main/scala/com/choreograph/tyda/Num.scala:26]: I wonder if we should make this a pure marker trait and move the implementations to ExprEvaluation since that the only place it is needed.

E.g we have other traits like `Groupable`, `Orderable`, `CanCast` that are pure markers and then `ExprEvaluation` is the only place that define the expected behavior.

File: tyda/src/test/scala/com/choreograph/tyda/ExprSpec.scala
Comment (eejbyfeldt) [inline tyda/src/test/scala/com/choreograph/tyda/ExprSpec.scala:1]: Seems like we have this block

```
  test("support int numeric operators") {
    assertBinaryNumericExpr[Int](_ + _, (l, r) => ExprNode.Add(Num[Int], l, r))
    assertBinaryNumericExpr[Int](_ - _, (l, r) => ExprNode.Subtract(Num[Int], l, r))
    assertBinaryNumericExpr[Int](_ * _, (l, r) => ExprNode.Multiply(Num[Int], l, r))
    assertBinaryNumericExpr[Int](_ / _, (l, r) => ExprNode.Quotient(Num[Int], l, r))
    assertUnaryNumericExpr[Int](-_, operand => ExprNode.Negate(Num[Int], operand))
  }
```

repeated for several types, so we should probably move it to a helper method and only have it once.


But I also if there specs are really needed and worth the potential maintenance burden. We are already going to add the behavioral specs that also covers that public api is usable as expected. So what do we additionally get here?

In general tests that repeats the internal ast structure have quite high cost as we have and probably will in the future change the ast structure.

File: tyda/src/main/scala/com/choreograph/tyda/ExprApi.scala [hunk 9]
```diff
@@ -502,26 +502,43 @@ trait ExprApi[Expr[T]] {
...
-  }
+      lift(ExprNode.Add(Num[T], unlift(lhs), unlift(AsExpr(rhs))))
 
-  extension [T: Integral](lhs: Expr[T]) {
+    /** Returns the difference of two expressions.
+      *
+      * Throws an exception if the operation leads to overflow.
+      */
+    infix def -[I: AsExpr.Of[T]](rhs: I): Expr[T] =
+      lift(ExprNode.Subtract(Num[T], unlift(lhs), unlift(AsExpr(rhs))))
```
Comment (eejbyfeldt) [inline tyda/src/main/scala/com/choreograph/tyda/ExprApi.scala]: Since the AsExpr conversion is only used to support literals `T` and `Expr[T]` here I would vote for defining this as

```suggestion
    infix def -(rhs: Expr[T]): Expr[T] =
      lift(ExprNode.Subtract(Num[T], unlift(lhs), unlift(AsExpr(rhs))))
      
    infix def -(rhs: T): Expr[T] = lhs - lit(rhs)
```

That gives much better error messages over depending on the AsExpr conversion.

File: tyda-spark/src/main/scala/com/choreograph/tyda/spark/ExprOnSpark.scala
Comment (eejbyfeldt) [inline tyda-spark/src/main/scala/com/choreograph/tyda/spark/ExprOnSpark.scala:1]: Feels like some logic here is duplicated between here and the ExprUnparser. Would it be possible to implement the logic using the ExprNodes instead have have it be shared?

File: tyda-sql/src/test/resources/golden/ExprEvaluationSparkSqlGoldenSuite.golden [hunk 10]
```diff
@@ -82,11 +82,27 @@ SELECT unix_micros(t1.value) AS value FROM t1
 SELECT timestamp_micros(unix_micros(t1.value)) AS value FROM t1
 ------ end
 
+------ Test: add two byte numbers
+SELECT (t1._1 + t1._2) AS value FROM t1
+------ end
+
+------ Test: add two double numbers
+SELECT CASE WHEN (((NOT ((t1._1 = CAST('Infinity' AS DOUBLE)) OR (t1._1 = CAST('-Infinity' AS DOUBLE)))) AND (NOT ((t1._2 = CAST('Infinity' AS DOUBLE)) OR (t1._2 = CAST('-Infinity' AS DOUBLE))))) AND ((NOT isnan((t1._1 + t1._2))) AND (((t1._1 + t1._2) < CAST(-1.7976931348623157E308 AS DOUBLE)) OR ((t1._1 + t1._2) > CAST(1.7976931348623157E308 AS DOUBLE))))) THEN raise_error('Double overflow') ELSE (t1._1 + t1._2) END AS value FROM t1
```
Comment (eejbyfeldt) [inline tyda-sql/src/test/resources/golden/ExprEvaluationSparkSqlGoldenSuite.golden]: I think we should try and quantify how much slower this is than the builtin `+` operator. If it so much slower that it not usable in a lot of contexts I think we might need to consider alternatives on how to handle this.

Maybe we could have a concept of some edge cases being implementation defined behavior.

**Review by eejbyfeldt — COMMENTED** [review]
Can we add a description with implementation choices made here and their motivations.

File: tyda-sql/src/test/resources/golden/ExprEvaluationSparkSqlGoldenSuite.golden (hunk 4 above)
Comment (NBorge) [reply tyda-sql/src/test/resources/golden/ExprEvaluationSparkSqlGoldenSuite.golden:117]: I've anchored the behaviour to the BigQuery implementation, which seems to be pretty close to the ANSI SQL standard. The chosen behaviour is as follows.
1. Division by zero results in an error for all types.
2. Arithmetic leading to overflow is an error for all types.
2.1 Arithmetic with values that are already Inf is allowed.

Where 2.1 seems to be the common behaviour while being an extension to the standard. The reason for the complicated parts of the BigQuery code is that floats aren't supported. See for example the equivalent BigQuery test for adding two doubles.

https://github.com/unum-io/tyda/pull/112/changes/BASE..661eb25945247be89de208fc884abf12f251988b#diff-fab06b79949b6eb66de7e7e70a85d0d224eadde192202fe4a9d443ca41cbc801R89-R91

File: tyda-sql/src/main/scala/com/choreograph/tyda/sql/ExprUnparser.scala [hunk 11]
```diff
@@ -155,6 +156,96 @@ private def exprToSqlExpr[T](fullExpr: ExprNode[T], args: UnparserArgs): Result[
...
+    ): SqlExpr = {
+      val overflow = SqlExpr.BinaryOp(
+        "AND",
+        SqlExpr.not(SqlExpr.Function(dialect.isNanFunction, Seq(checkedResult))),
+        SqlExpr.BinaryOp(
+          "OR",
+          SqlExpr.BinaryOp("<", checkedResult, min),
+          SqlExpr.BinaryOp(">", checkedResult, max)
+        )
+      )
```
Comment (eejbyfeldt) [inline tyda-sql/src/main/scala/com/choreograph/tyda/sql/ExprUnparser.scala]: Because all comparisons with nan being false I think the `NaN` check is not needed here. 
```suggestion
      val overflow = SqlExpr.BinaryOp(
          "OR",
          SqlExpr.BinaryOp("<", checkedResult, min),
          SqlExpr.BinaryOp(">", checkedResult, max)
        )
```

File: tyda-spark/src/main/scala/com/choreograph/tyda/spark/ExprOnSpark.scala [hunk 12]
```diff
@@ -119,6 +119,35 @@ private class ExprOnSpark[T](cfs: Map[ExprNode.Reference[?], ColumnFactory[?]])
...
+      lhs: Column,
+      rhs: Column,
+      checkedResult: Column,
+      returnedResult: Column,
+      codec: Codec[?]
+  ): Column =
+    codec match {
+      case Codec.Float => when(
+          !isnan(checkedResult) && !isInfinite(lhs, codec) && !isInfinite(rhs, codec) &&
+            (checkedResult < lit(-Float.MaxValue) || checkedResult > lit(Float.MaxValue)),
```
Comment (eejbyfeldt) [inline tyda-spark/src/main/scala/com/choreograph/tyda/spark/ExprOnSpark.scala]: If we use abs things becomes slightly more simpler here:

```suggestion
          !isnan(checkedResult) && !isInfinite(lhs, codec) && !isInfinite(rhs, codec) &&
            (abs(checkedResult) > lit(Float.MaxValue)),
```

File: tyda-spark/src/main/scala/com/choreograph/tyda/spark/ExprOnSpark.scala [hunk 13]
```diff
@@ -119,6 +119,35 @@ private class ExprOnSpark[T](cfs: Map[ExprNode.Reference[?], ColumnFactory[?]])
   private def cfFromRef(ref: ExprNode.Reference[?]): ColumnFactory[?] =
     cfs.get(ref).getOrElse(Errors.failUnexpectedReference(ref, cfs.keys))
 
+  private def isInfinite(column: Column, codec: Codec[?]): Column =
+    codec match {
+      case Codec.Float => column === lit(Float.PositiveInfinity) || column === lit(Float.NegativeInfinity)
+      case Codec.Double => column === lit(Double.PositiveInfinity) || column === lit(Double.NegativeInfinity)
```
Comment (eejbyfeldt) [inline tyda-spark/src/main/scala/com/choreograph/tyda/spark/ExprOnSpark.scala]: Should we implement this as 
```suggestion
      case Codec.Float => abs(column) > lit(Float.MaxValue)
      case Codec.Double => abs(column) > lit(Double.MaxValue)
```

since that has fewer operations and matches how it is implemented on the jvm https://github.com/openjdk/jdk/blob/26bf445f4726f1936a0a4cbaf1424c5235424bfb/src/java.base/share/classes/java/lang/Float.java#L630

File: tyda-sql/src/test/resources/golden/ExprEvaluationSparkSqlGoldenSuite.golden (hunk 4 above)
Comment (eejbyfeldt) [reply tyda-sql/src/test/resources/golden/ExprEvaluationSparkSqlGoldenSuite.golden:117]: That make sense. Lets put the motivation in the description of the PR.

File: tyda-spark/src/main/scala/com/choreograph/tyda/spark/ExprOnSpark.scala
Comment (eejbyfeldt) [reply tyda-spark/src/main/scala/com/choreograph/tyda/spark/ExprOnSpark.scala:1]: We have other shared logic in the tyda-rewrite module.

File: tyda/src/main/scala/com/choreograph/tyda/ExprApi.scala (hunk 9 above)
Comment (NBorge) [reply tyda/src/main/scala/com/choreograph/tyda/ExprApi.scala]: Added your suggestion with slight changes, e.g.:
```
infix def +(rhs: Expr[T]): Expr[T] = lift(ExprNode.Add(Num[T], unlift(lhs), unlift(rhs)))
infix def +(rhs: T)(using Codec[T]): Expr[T] = lhs + lit(rhs)
```

Comment (eejbyfeldt) [reply tyda/src/main/scala/com/choreograph/tyda/ExprApi.scala]: Could we take the codec from the lhs instead? e.g
```
infix def +(rhs: Expr[T]): Expr[T] = lift(ExprNode.Add(Num[T], unlift(lhs), unlift(rhs)))
infix def +(rhs: T): Expr[T] = lhs + lit(rhs)(using unlift(lhs).codec)
```

Comment (eejbyfeldt) [reply tyda/src/main/scala/com/choreograph/tyda/ExprApi.scala]: There is some compile time cost with using givens. So when it possible to avoid it by doing slightly more "work" internally we try to do that.

File: tyda-sql/src/test/resources/golden/ExprEvaluationBigQueryGoldenSuite.golden [hunk 14]
```diff
@@ -82,14 +82,66 @@ SELECT unix_micros(t1.value) AS value FROM t1
...
+SELECT (t1._1 + t1._2) AS value FROM t1
+------ end
+
+------ Test: add two double numbers
+SELECT (t1._1 + t1._2) AS value FROM t1
+------ end
+
+------ Test: add two float numbers
+SELECT WITH(c0 AS t1._1, WITH(c1 AS t1._2, WITH(c2 AS (c0 + c1), CASE WHEN ((((NOT (abs(c0) = CAST('Infinity' AS FLOAT64))) AND (NOT (abs(c1) = CAST('Infinity' AS FLOAT64)))) AND (NOT is_nan(c2))) AND (NOT (abs(c2) <= CAST(3.4028234663852886E38 AS FLOAT64)))) THEN CAST(error('Float overflow') AS FLOAT64) ELSE c2 END))) AS value FROM t1
+------ end
```
Comment (dahlbaek) [inline tyda-sql/src/test/resources/golden/ExprEvaluationBigQueryGoldenSuite.golden:119]: I seem to be getting reasonable results when I do something like

```
bq query --use_legacy_sql=false '
CREATE TEMP FUNCTION add_float(x FLOAT64, y FLOAT64)
RETURNS FLOAT64
AS (
  CASE
  WHEN (ABS(x) > CAST(3.4028234663852886E38 AS FLOAT64))
    OR (ABS(y) > CAST(3.4028234663852886E38 AS FLOAT64))
    OR IS_NAN(x+y)
    OR (ABS(x+y) > CAST(3.4028234663852886E38 AS FLOAT64))
  THEN CAST(ERROR('"'"'Float overflow'"'"') AS FLOAT64)
  ELSE x+y
  END
);

SELECT a, b, c, add_float(a, b) `a+b`, add_float(a,add_float(b,c)) AS `a+b+c`
FROM UNNEST([
--  STRUCT(2.5 AS a, 3.5 AS b, CAST('"'"'Infinity'"'"' AS FLOAT64) as c),
--  STRUCT(2.5 AS a, CAST('"'"'Infinity'"'"' AS FLOAT64) AS b, 3.5 as c),
--  STRUCT(CAST('"'"'Infinity'"'"' AS FLOAT64) AS a, 2.5 AS b, 3.5 as c),
  STRUCT(1.0 AS a, 2.0 AS b, 100.2 as c),
  STRUCT(100.25 AS a, 200.75 AS b, 0.1 as c),
  STRUCT(-10.0 AS a, 4.0 AS b, -2.3 as c)
]);
'
```

i.e. remove the `--` from any of the commented lines will cause failure. Am I missing something, or could this be a way to produce readable sql and avoid exponential increases in query size due to recursive inlining of the add_float body?

Comment (dahlbaek) [reply tyda-sql/src/test/resources/golden/ExprEvaluationBigQueryGoldenSuite.golden:119]: @NBorge if you agree that it would be reasonable to do something with temp functions, then I think we should park this one until we can merge https://github.com/unum-io/tyda/pull/163, and then use https://spark.apache.org/docs/latest/sql-ref-syntax-ddl-create-sql-function.html to have unified approach between spark and bigquery.

Comment (NBorge) [reply tyda-sql/src/test/resources/golden/ExprEvaluationBigQueryGoldenSuite.golden:119]: Yes, let's park this and try this after the merge. This should be possible, and I think quite readable.
